src/utils/hierarchical_model.py

```diff
@@ -0,0 +1,213 @@
+        method_type (str, optional): The clustering algorithm to use. Default is 'ward'.
+        distance_metric (str, optional): The distance metric to use. Default is 'euclidean'.
+        optimal_ordering (bool, optional): Whether to compute the optimal leaf ordering for the dendrogram. Default is False.
+
+    Returns:
+        linkage_matrix (numpy.ndarray): The hierarchical clustering linkage matrix.
+    """
+    return linkage(data, method=method_type, metric=distance_metric, optimal_ordering=optimal_ordering)
+
+
+def plot_dendrogram_all_methods(data: pd.DataFrame, method_list=None, distance_metric='euclidean', optimal_ordering=False) -> None:
+    """Calculates the linkage matrix and display the dendrogram plot for each given linkage method.
+
+    Args:
+        data (pd.DataFrame): The dataset to draw the dendrogram plot of.
+        method_list (list): The linkage methods to use. Defaults to ['single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward'].
+        distance_metric (str): The distance metric to use when calculating the linkage matrix. Defaults to 'euclidean'.
+        optimal_ordering (bool): Whether to use optimal leaf ordering when drawing the dendrogram plot. Defaults to False.
+    """
+
+    # Set default linkage methods if not provided
+    if method_list is None:
+        method_list = ['single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward']
+
+    # Create a figure to plot the dendrograms
+    fig, axs = plt.subplots(len(method_list), 1, figsize=(10, 30), sharex=True, gridspec_kw={'hspace': 0.5})
+
+    # Loop over the methods and plot the corresponding dendrogram
+    for index, method in enumerate(method_list):
+        # Calculate the linkage matrix
+        linkage_matrix = get_linkage(data, method, distance_metric, optimal_ordering)
+        
+        # Plot the dendrogram
+        dendrogram(linkage_matrix, ax=axs[index])
+        
+        # Set the plot title and axis labels
+        axs[index].set_title(f'Dendrogram ({method} linkage)')
+        axs[index].set_xlabel('Sample Index')
+        axs[index].set_ylabel('Distance')
+    
+    # Show the plot and return the dendrogram object
+    plt.show()
+
+
+def plot_thresholded_dendrogram(linkage_matrix: ndarray, distance_threshold: int) -> None:
+    """Plots the dendrogram with a horizontal line at the given threshold.
+
+    Args:
+        linkage_matrix (numpy.ndarray): The linkage matrix of the dataset.
+        threshold (float): The threshold distance to plot the horizontal line at.
+    """
+    # Plot the dendrogram
+    plt.figure(figsize=(10, 7))
+    dendrogram(linkage_matrix)
+    
+    # Set the plot title and axis labels
+    plt.title("Dendrogram with Threshold")
+    
+    # Draw the horizontal line
+    plt.axhline(y=distance_threshold, color='black', linestyle='--')
+    
+    # Show the plot
+    plt.show()
+
+
+def hierarchical_clustering(data: pd.DataFrame, k=3) -> ndarray:
```

Comment:
No need for this function.

Reply:
I use this function in several places in the notebooks.

Reply:
Yes but the function only returns a value. you can do this without initializing a function so why did you use it?

Reply:
It's easier...